platform/darwin/src/collator.mm

```diff
@@ -35,7 +35,13 @@ int compare(const std::string& lhs, const std::string& rhs) const {
     }
 
     std::string resolvedLocale() const {
-        return [locale localeIdentifier].UTF8String;
+        // Following documentation at:
+        // https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15-SW9
+        // We expect NSLocale to accept BCP 47 tags as localeIdentifier inputs, but the output format
+        // may append the region tag with an "_". Changing that to a "-" makes the identifier BCP 47 compliant.
+        // Experimentally, "zh-Hans-HK" and "zh-Hans_HK" both round trip -- if the second is used by
+        //  `currentLocale`, we don't want to return the underscore.
+        return [[locale localeIdentifier] stringByReplacingOccurrencesOfString:@"_" withString:@"-"].UTF8String;
```

Comment:
Yep, this is good enough to avoid the underscore issue, since none of the ISO or UN region codes have hyphens or underscores. If we want to be super safe and protect against extensions in the locale identifier, then this code would ensure that we only get the bits we care about:

```objc
NSMutableString *resolvedLocale = [NSMutableString string];
// NSLocale.languageCode & company are unavailable in iOS 9.
NSString *languageCode = [locale objectForKey:NSLocaleLanguageCode];
if (languageCode) {
    [resolvedLocale appendFormat:@"-%@", languageCode];
}
NSString *scriptCode = [locale objectForKey:NSLocaleScriptCode];
if (scriptCode) {
    [resolvedLocale appendFormat:@"-%@", scriptCode];
}
NSString *regionCode = [locale objectForKey:NSLocaleCountryCode];
if (regionCode) {
    [resolvedLocale appendFormat:@"-%@", regionCode];
}
return resolvedLocale.UTF8String;
```

(Or the C++ string stream equivalent.) But I’m not even sure that an extension can ever be present in `+[NSLocale currentLocale]`. For example, my Mac’s locale settings are all over the place, including custom sorting and date formats, but the locale identifier is still just `vi_US`, which would be fine with just the underscore replacement.